src/Features/Core/Portable/Completion/Providers/AbstractEnumCompletionProvider.cs

```diff
@@ -0,0 +1,142 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Recommendations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractEnumCompletionProvider : AbstractSymbolCompletionProvider
+    {
+        private static readonly CompletionItemRules s_rules = CompletionItemRules.Default.WithMatchPriority(MatchPriority.Preselect);
+
+        protected abstract (string displayText, string suffix, string insertionText) GetDefaultDisplayAndSuffixAndInsertionText(ISymbol symbol, SyntaxContext context);
+
+        protected override Task<ImmutableArray<ISymbol>> GetPreselectedSymbolsAsync(SyntaxContext context, int position, OptionSet options, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = context.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFacts.IsInNonUserCode(context.SyntaxTree, context.Position, cancellationToken))
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            // This providers provides fully qualified names, eg "DayOfWeek.Monday"
+            // Don't run after dot because SymbolCompletionProvider will provide
+            // members in situations like Dim x = DayOfWeek.$$
+            if (context.TargetToken.RawKind == syntaxFacts.SyntaxKinds.DotToken)
+                return SpecializedTasks.EmptyImmutableArray<ISymbol>();
+
+            var typeInferenceService = context.GetLanguageService<ITypeInferenceService>();
+            var inferedTypes = typeInferenceService.InferTypes(context.SemanticModel, position, cancellationToken);
```

Comment:
I changed this call from `InferType` to `InferTypes` as there might be more than one Enum suitable at the position (e.g. method overloads:

https://github.com/dotnet/roslyn/blob/97559e3bd9b136cb2e5f4797b860580f03ae99fd/src/EditorFeatures/CSharpTest/Completion/CompletionProviders/EnumCompletionProviderTests.cs#L567-L598

Reply:
that seemes reasonable.